OOXML Relationships Transform support

vmiklos opened the pull request.

vmiklos commented:
One more detail I forgot to mention: it's important that the hash in the testcase taken from the reference implementation (Microsoft Office), so in case it would not match (it does), then the code needs fixing, not the hash value. :-)

lsh123 commented on src/xpath.c:
```diff
@@ -1163,3 +1164,553 @@ xmlSecTransformVisa3DHackExecute(xmlSecTransformPtr transform, int last,
+
+    cur = node->children;
+    while(cur != NULL) {
+        if(xmlSecCheckNodeName(cur, "RelationshipReference", "http://schemas.openxmlformats.org/package/2006/digital-signature")) {
```
Could you please put these string constants into strings.[h|c]?

lsh123 commented on src/xpath.c:
```diff
@@ -1163,3 +1164,553 @@ xmlSecTransformVisa3DHackExecute(xmlSecTransformPtr transform, int last,
+    id1 = xmlGetProp(node1, "Id");
+    id2 = xmlGetProp(node2, "Id");
+
+    if(id1 == NULL)
```
Code style: braces are required

lsh123 commented on src/xpath.c:
```diff
@@ -1163,3 +1164,553 @@ xmlSecTransformVisa3DHackExecute(xmlSecTransformPtr transform, int last,
+
+    switch(cur->type) {
+    case XML_ELEMENT_NODE:
+        if(xmlSecCheckNodeName(cur, "Relationship", "http://schemas.openxmlformats.org/package/2006/relationships")) {
```
Strings should be moved to strings.[h|c]

lsh123 commented on src/xpath.c:
```diff
@@ -1163,3 +1164,553 @@ xmlSecTransformVisa3DHackExecute(xmlSecTransformPtr transform, int last,
+    case XML_ELEMENT_NODE:
+        if(xmlSecCheckNodeName(cur, "Relationship", "http://schemas.openxmlformats.org/package/2006/relationships")) {
+            xmlChar* id = xmlGetProp(cur, "Id");
+            if(id != NULL) {
```
Is the Id required? Is the Id always Id or it might be any 'id' attribute defined in schema?

lsh123 commented on src/xpath.c:
```diff
@@ -1163,3 +1164,553 @@ xmlSecTransformVisa3DHackExecute(xmlSecTransformPtr transform, int last,
+
+                ctx = xmlSecRelationshipGetCtx(transform);
+                for(i = 0; i < xmlSecPtrListGetSize(ctx->sourceIdList); ++i) {
+                    if(strcmp(xmlSecPtrListGetItem(ctx->sourceIdList, i), id) == 0) {
```
strcmp -> xmlStrcmp

lsh123 commented on src/xpath.c:
```diff
@@ -1163,3 +1164,553 @@ xmlSecTransformVisa3DHackExecute(xmlSecTransformPtr transform, int last,
+                    }
+                }
+
+                if(!found) {
```
Bug: found can be 0 (if the element is at 0 pos). Better initialize to -1 and check found < 0

lsh123 commented on src/xpath.c:
```diff
@@ -1163,3 +1164,553 @@ xmlSecTransformVisa3DHackExecute(xmlSecTransformPtr transform, int last,
+xmlSecTransformRelationshipProcessNode(xmlSecTransformPtr transform, xmlOutputBufferPtr buf, xmlNodePtr cur) {
+    int ret;
+
+    switch(cur->type) {
```
I don't understand why is this switch() needed? Doesn't look like it's doing anything... Also consider using xmlSecFindNode() function here to skip through text nodes (indentation, etc)

lsh123 commented on src/xpath.c:
```diff
@@ -1163,3 +1164,553 @@ xmlSecTransformVisa3DHackExecute(xmlSecTransformPtr transform, int last,
+    }
+
+    for(; cur; cur = cur->next) {
+        if(strcmp(cur->name, "Relationship") == 0) {
```
strcmp -> xmlStrcmp

lsh123 commented on src/xpath.c:
```diff
@@ -1163,3 +1164,553 @@ xmlSecTransformVisa3DHackExecute(xmlSecTransformPtr transform, int last,
+
+    for(; cur; cur = cur->next) {
+        if(strcmp(cur->name, "Relationship") == 0) {
+            xmlListInsert(list, cur);
```
Return value check is needed

lsh123 commented on src/xpath.c:
```diff
@@ -1163,3 +1164,553 @@ xmlSecTransformVisa3DHackExecute(xmlSecTransformPtr transform, int last,
+    xmlAttrPtr attr;
+    int foundTargetMode = 0;
+
+    xmlOutputBufferWriteString(buf, "<");
```
Retrun values checks are needed

lsh123 commented on src/xpath.c:
```diff
@@ -1163,3 +1164,553 @@ xmlSecTransformVisa3DHackExecute(xmlSecTransformPtr transform, int last,
+        xmlOutputBufferWriteString(buf, "\"");
+    }
+
+    if(strcmp(cur->name, "Relationship") == 0 && !foundTargetMode) {
```
strcmp -> xmlStrcmp

lsh123 commented on src/xpath.c:
```diff
@@ -1163,3 +1164,553 @@ xmlSecTransformVisa3DHackExecute(xmlSecTransformPtr transform, int last,
+       if(outSize > 0) {
+           xmlSecAssert2(xmlSecBufferGetData(&(transform->outBuf)), -1);
+
+           memcpy(data, xmlSecBufferGetData(&(transform->outBuf)), outSize);
```
how do we know that transform->outBuf has at least outSize data?

vmiklos commented:
When it's trivial address the problem (like strcmp -> xmlStrcmp), what do you prefer, pushing a follow-up commit or a new version of the original commit? The later sounds cleaner, but I'm not sure if github then throws away your comments or not... :-)

lsh123 commented:
You can keep updating the patch. It will remove comments if the line was changed

vmiklos commented:
Changes in v2:

Changed the testcase to use an external reference, as that's what gets used in
practice, so more relevant.

> Could you please put these string constants into strings.[h|c]?

Done.

> Code style: braces are required

Done.

> Is the Id required? Is the Id always Id or it might be any 'id' attribute defined in schema?

Yes, it's required and it should be Id. I've changed the code to error out if
it's missing.

> strcmp -> xmlStrcmp

Done.

> Bug: found can be 0 (if the element is at 0 pos). Better initialize to -1 and check found < 0

I use `found = 1`, not `found = i`, so I don't think it's a bug, but sure, -1
as default can't hurt -- changed.

> I don't understand why is this switch() needed? Doesn't look like it's doing
> anything... Also consider using xmlSecFindNode() function here to skip
> through text nodes (indentation, etc)

It's indeed not needed, removed.

I think I don't need xmlSecFindNode() here, as the purpose of
xmlSecTransformRelationshipProcessNode() is to filter out <Relationship>
elements with an Id attribute that's not whitelisted in the transformation
argument list, and invoke xmlSecTransformRelationshipProcessElementNode() for
the rest of the elements. In practice if the input conforms to the schema, then
it won't contain any text node. Unless I miss something about how
xmlSecFindNode() could simplify code here. :-)

> Return value check is needed [ for xmlListInsert ]

Done.

> Retrun values checks are needed [ for xmlOutputBufferWriteString ]

Done.

> how do we know that transform->outBuf has at least outSize data?

Code above memcpy() does this:

```
out = &transform->outBuf;
buf = xmlSecBufferCreateOutputBuffer(out);
outSize = xmlSecBufferGetSize(out);
```

That makes me believe know that. Or did I miss something?

lsh123 commented:
I've made a few more changes:
1) moved this transform into its own file
2) fixed a few pedantic compile errors
3) fixed some style issue.

"Code above memcpy() does this:"
I've cleaned up the code a little to make it easier to understand. You are correct there is 

I had to create a new pull request: https://github.com/lsh123/xmlsec/pull/24/files

Could you please take a look?

lsh123 commented:
Also, could you please help me to map the steps/sub-steps from the transform description to the code? May be you can comment on the new PR on the lines that are relevant to the steps?
